File: 2019/go/day14.go
package main

import (
	"AdventOfCode/util"
	"fmt"
	"math"
	"strconv"
	"strings"
)

type reagent struct {
	name string
	qty  int
}

type reaction struct {
	qty      int
	reagents []reagent
}

var reactions map[string]reaction
var chemicals []string

func parseChemical(s string) (int, string) {
	values := strings.Split(s, " ")
	if n, err := strconv.Atoi(values[0]); err == nil {
		return n, values[1]
	}
	return 0, ""
}

func parse(lines []string) map[string]reaction {
	reactions := make(map[string]reaction)
	for _, l := range lines {
		sides := strings.Split(l, " => ")

		// inputs
		inputs := strings.Split(sides[0], ", ")
		reagents := make([]reagent, len(inputs))
		for i, input := range inputs {
			qty, chem := parseChemical(input)
			reagents[i].name = chem
			reagents[i].qty = qty
		}

		// output
		qty, chem := parseChemical(sides[1])
		reactions[chem] = reaction{
			qty,
			reagents,
		}
	}
	return reactions
}

func getChemicalsInOrder() []string {
	var aux func(string)
	var result []string

	visited := make(map[string]bool)
	aux = func(c string) {
		visited[c] = true
		for _, reagent := range reactions[c].reagents {
			if !visited[reagent.name] {
				aux(reagent.name)
			}
		}
		result = append(result, c)
	}

	for chem := range reactions {
		if !visited[chem] {
			aux(chem)
		}
	}
	return result
}

func getOreQty(fuel int) int {
	required := make(map[string]int)
	required["FUEL"] = fuel
	for i := len(chemicals) - 1; i >= 0; i-- {
		chem := chemicals[i]
		req := required[chem]
		reaction := reactions[chem]
		m := int(math.Ceil(float64(req) / float64(reaction.qty)))
		for _, reagent := range reaction.reagents {
			required[reagent.name] += m * reagent.qty
		}
	}
	return required["ORE"]
}

func part1() int {
	return getOreQty(1)
}

func part2() int {
	min := 1e12 / getOreQty(1)
	max := 2 * min

	for min < max {
		mid := (max+min)/2 + 1
		ore := getOreQty(mid)
		if ore > 1e12 {
			max = mid - 1
		} else {
			min = mid
		}
	}
	return min
}

func main() {
	input, _ := util.ReadInput("../day_14.txt")
	reactions = parse(input)
	chemicals = getChemicalsInOrder()

	fmt.Printf("part 1: %d\n", part1())
	fmt.Printf("part 2: %d\n", part2())
}

File: 2019/python/day14.py
from collections import defaultdict
from math import ceil

def parse(s):
    q, m = s.split()
    return (int(q), m)

def parse_equations(lines):
    e = {}
    for line in lines:
        left, right = line.rstrip().split(' => ')
        amount, mat = parse(right)
        e[mat] = (amount, [parse(s) for s in left.split(', ')])
    return e

def calc_ore_amount(amount):
    inventory = defaultdict(int, {'FUEL': -amount})
    work = ['FUEL']
    while work:
        material = work.pop(0)
        quantity = -inventory[material]
        produced, inputs = equations[material]
        mult = ceil(quantity / produced)
        inventory[material] += mult * produced
        for n, mat in inputs:
            inventory[mat] -= mult * n
            if inventory[mat] < 0 and mat != 'ORE':
                work.append(mat)

    return -inventory['ORE']


with open('../day_14.txt') as f:
    equations = parse_equations(f.readlines())
ore = calc_ore_amount(1)
print('part 1:', ore)


min_fuel = 10**12 // ore
max_fuel = 2 * min_fuel

while min_fuel < max_fuel:
    fuel = (max_fuel + min_fuel) // 2 + 1
    ore = calc_ore_amount(fuel)
    if ore > 10**12:
        max_fuel = fuel - 1
    else:
        min_fuel = fuel
print('part 2:', max_fuel)

